include/ocpp/common/message_queue.hpp

```diff
@@ -342,44 +345,52 @@ template <typename M> class MessageQueue {
+                        continue;
+                    }
+
+                    message = *normal_message_it;
+                    queue_type = QueueType::Normal;
+                    selected_normal_message_it = normal_message_it; // Save iterator to the selected message
+                    break;                                          // Stop searching once a sendable message is found
                 }
 
-                if (!this->transaction_message_queue.empty()) {
-                    auto& transaction_message = this->transaction_message_queue.front();
-                    EVLOG_debug << "transaction msg timestamp: " << transaction_message->timestamp;
-                    if (message == nullptr) {
-                        if (transaction_message->timestamp <= now) {
-                            EVLOG_debug << "transaction message timestamp <= now";
-                            message = transaction_message;
-                            queue_type = QueueType::Transaction;
-                        }
-                    } else {
-                        if (transaction_message->timestamp <= message->timestamp and
-                            message->messageType != M::BootNotification) {
-                            EVLOG_debug << "transaction message timestamp <= normal message timestamp";
-                            message = transaction_message;
-                            queue_type = QueueType::Transaction;
-                        } else {
-                            EVLOG_debug << "Prioritizing newer normal message over older transaction message";
-                        }
+                auto selected_transaction_message_it = transaction_message_queue.end();
+                for (auto transaction_message_it = transaction_message_queue.begin();
+                     transaction_message_it != transaction_message_queue.end(); ++transaction_message_it) {
+
+                    if ((*transaction_message_it)->stall_until_accepted and !this->is_registration_status_accepted) {
+                        continue;
                     }
+
+                    if ((*transaction_message_it)->timestamp > now) {
+                        continue;
+                    }
+
+                    if (message != nullptr and ((*transaction_message_it)->timestamp > message->timestamp or
+                                                message->messageType == M::BootNotification)) {
+                        continue;
+                    }
+
+                    message = *transaction_message_it;
+                    queue_type = QueueType::Transaction;
+                    selected_transaction_message_it = transaction_message_it; // Save iterator to the selected message
+                    break; // Stop searching once a sendable message is found
                 }
 
                 if (message == nullptr) {
                     EVLOG_debug << "No message in queue ready to be sent yet";
+                    this->new_message = false;
```

Comment:
See https://github.com/EVerest/libocpp/blob/main/include/ocpp/common/message_queue.hpp#L449 : `new_message` was set to false only when `transaction_message_queue` and `normal_message_queue` is empty to allow the cv to wait for the next notify when a new message is added to the queue.

With the changes added we can have a situation where there are messages on the queue, but we dont want to process the messages on the queue because its not allowed to send them, so we want to wait until the next message until we continue processing.